File: golang/validParenthesis.go
package main

func IsValidParenthesis(s string) bool {
	map1 := map[byte]byte{
		')': '(',
		'}': '{',
		']': '[',
	}

	stack := []byte{}
	for i := 0; i < len(s); i++ {
		char := s[i]
		if expected, ok := map1[char]; ok {
			top := stack[len(stack)-1]
			if top != expected {
				return false
			}
			stack = stack[:len(stack)-1] // pop
		} else {
			stack = append(stack, char)
		}
	}
	return len(stack) == 0
}

File: typescript/validParenthesis.ts
function isValidParenthesis(s: string): boolean {
 const pairs: Record<string, string> = {
   ")": "(",
   "}": "{",
   "]": "["
 }

 const stack: string[] = []
 for (const char of s) {
   if (char in pairs) {
     // closing bracket → check match
     const expected = pairs[char]
     const top = stack.pop();
     if (top !== expected) {
       return false;
     }
   } else {
     // opening bracket → push to stack
     stack.push(char);
   }
 }
 return stack.length === 0
}

console.log(isValidParenthesis("({[]})"))
